Implement backend for /title command

scottwillmoore opened the pull request:
My current progress of the implementation of the /title command (#1806), also noticed while implementing it that the document is not in complete alphabetical order as the command `virtual void SendDisplayObjective` appears near the bottom.

scottwillmoore commented:
The integration tests have failed, so I will have a look at fixing those up tomorrow. Is this because I have used tabs for the indentation of my code? I thought in the CONTRIBUTING guide it mentions to use tabs for indentation? It also says I need two spaces in front of a style comment? `// comment >> //  comment`, however many of the comments I just copied and pasted from other sections of the source such as `// In game mode?`.

NiLSPACE commented on src/Protocol/Protocol.h:
```diff
@@ -86,6 +86,7 @@ class cProtocol
 	virtual void SendExplosion                  (double a_BlockX, double a_BlockY, double a_BlockZ, float a_Radius, const cVector3iArray & a_BlocksAffected, const Vector3d & a_PlayerMotion) = 0;
 	virtual void SendGameMode                   (eGameMode a_GameMode) = 0;
 	virtual void SendHealth                     (void) = 0;
+	virtual void SendHideTitle					(void) = 0;
```
Only the at the start of a line you align the code using tabs. After that you use spaces.

NiLSPACE commented on src/Protocol/Protocol18x.cpp:
```diff
@@ -542,6 +542,18 @@ void cProtocol180::SendHealth(void)
 
+void cProtocol180::SendHideTitle(void)
+{
+	ASSERT(m_State == 3); // In game mode?
```
You need to have 2 spaces after `//`. That's why the build is failing.

scottwillmoore commented:
Yeah just looking through some of it now, sorry about that I will fix it up tomorrow.

scottwillmoore commented:
Ah I see thanks, you can tell it is late because I thought CheckBasicStyle.lua was used to check lua styles! That will help heaps, thanks for that. Decided to fix it up tonight, enjoy. :)

madmaxoft commented on src/Protocol/Protocol18x.cpp:
```diff
@@ -1123,6 +1147,68 @@ void cProtocol180::SendDisplayObjective(const AString & a_Objective, cScoreboard
+	Pkt.WriteVarInt32(1);  // Set action as, set subtitle
+
+	Pkt.WriteString(Printf("{\"text\":\"%s\"}", EscapeString(a_SubTitle).c_str()));
+}
```
We might want to rename this to `SendSetRawSubtitle()` and send the input string as-is, so that the `/title` command can send the entire string through here unparsed unmodified.

madmaxoft commented on src/Protocol/Protocol18x.cpp:
```diff
@@ -542,6 +542,18 @@ void cProtocol180::SendHealth(void)
+	ASSERT(m_State == 3);  // In game mode?
+
+	cPacketizer Pkt(*this, 0x45);  // Title packet
+	Pkt.WriteVarInt32(3);  // Set action as, hide title
```
The comment is very weird, why the comma in the middle?

scottwillmoore commented:
Hello, yep I agree with the name changes, they make the function easier to understand, and also you specified in the issue that it was `SendSetSubTitle(AString & a_SubTitle)` however now you are using `SendSetSubtitle(AString & a_Subtitle)`? Would you like me to change to using the second form? Subtitle is an actual word so it would make sense to display it the second way in the functions.

With the actual functionality of the `SendSetRawSubtitle(AString & a_Subtitle)` is there a reason you would want to send the string unformatted with json, as I am pretty sure if the text is not in the json format it will not work when it reaches the client. The `cCompositeChat` function allows you to send a fully formatted json text to the client and I just viewed the `AString` version as a shorthand to sending an unformatted, plain text title which gets wrapped in a basic json structure, whithout requiring the user to create a `cCompositeChat` item. I guess an advantage of not wrapping the string in the json structure would be that users could send their own json string through the function, rather than sending a `cCompositeChat` structure through the other function? I doubt many people would do this however as it is probably easier to create a `cCompositeChat` object.

Ah, yeah the comment is a bit weird, I think I put the comma there to separate the `hide title` which is the action specified by the protocol, from the other text. I can probably just remove the comma or replace it with a colon, which would make more grammatical sense.

tigerw commented:
About refactoring: http://stackoverflow.com/questions/363292/why-is-visual-c-lacking-refactor-functionality

madmaxoft commented:
If you're asking about the uppercase `T` vs lowercase `t` in `SubTitle`, I don't really care about those. I care more about the `Raw` being present.

The point is, how will we implement the `title <player> subtitle <raw json title>` command handler? Without the `Raw` variant of the call, it would have to parse the raw json title into a `cCompositeChat` (possibly losing things) and then the call would re-serialize that into a json string? That seems somewhat unnecessary.

If we wanted to have a simple interface for sending one single string, we could do that on the `cClientHandle` level (which gets exported to Lua API) and just make it call the `Raw` cProtocol function.

madmaxoft commented on src/Protocol/Protocol18x.cpp:
```diff
@@ -1123,6 +1147,68 @@ void cProtocol180::SendDisplayObjective(const AString & a_Objective, cScoreboard
+	cPacketizer Pkt(*this, 0x45);  // Title packet
+	Pkt.WriteVarInt32(1);  // Set subtitle
+
+	Pkt.WriteString(a_SubTitle.CreateJsonString(ShouldUseChatPrefixes));
```
I think neither titles nor subtitles should ever have chat prefixes, or should they? @tigerw what do you think?
Also, how about just calling `SendSetRawSubTitle(a_SubTitle.CreateJsonString(false))`? That way the actual packet format is centralized to a single location and it's easier to clone+upgrade the protocol to a newer version.

tigerw replied:
I would agree that titles shouldn't be prefixed.

madmaxoft commented on src/Protocol/Protocol18x.cpp:
```diff
@@ -1123,6 +1147,68 @@ void cProtocol180::SendDisplayObjective(const AString & a_Objective, cScoreboard
+	cPacketizer Pkt(*this, 0x45);  // Title packet
+	Pkt.WriteVarInt32(0);  // Set title
+
+	Pkt.WriteString(a_Title.CreateJsonString(ShouldUseChatPrefixes));
```
Again, chat prefixes are questionable.

madmaxoft commented:
Sorry about the late reaction. Please add a comment to the PR when you think we should have a look, otherwise it doesn't pop up and we tend to forget :)

scottwillmoore commented:
Oh sorry about that. I wasn't aware that you wouldn't be notified if I didn't comment. Well I have fixed up the changes mentioned above and now `SendSetSubTitle(...)` and `SendSetTitle(...)` both use the raw implementation instead. Also removed chat prefixes.

madmaxoft commented:
This looks good now. Would you be willing to squash all the commits into one? We prefer not to have the "fix formatting" commits in the repo.

scottwillmoore commented:
Ok, you may have to check (I have never squashed commits into one before), I squashed them on my local repo which worked perfectly. I then went to push them to my github repo and it wouldn't let me. So I ended up using this command `git push origin +ImplementTitleCommand` which appears to have worked. On my github fork there is only one commit called `Implement backend for /title command` however in this pull request it still shows 4 commits in the commits tab above. Is this ok?

scottwillmoore commented:
Also we havn't tested to see if these functions will actually work? Is that what we will do now, so we can see if they work when sent to a client? Or does a lua wrapper for these functions need to be created so  the testing of these functions can be easily tested from a lua script.

madmaxoft commented:
That is correct, after squashing commits locally you need to force-push to the repo (because you're effectively changing the history). If multiple people are working on the repo, this gets dangerous, but right now it's only you so you're fine.

The GitHub page needs reloading, then it shows only the single commit.

In order to test the functions, you need to export them to the Lua API, then the easiest (and recommended) way to test is to write some test code in the Debuggers plugin. This will serve both as a test, and as example code.

To export a simple function to the API, you just need to add a `// tolua_export` comment after it in the header file. Namely, export the `cClientHandle` functions, as that's the proper interface to use them. When you build, you'll notice that ToLua is rebuilding the bindings as part of the build. Then you can use those functions from the plugins.

Howaner commented on src/ClientHandle.cpp:
```diff
@@ -2320,6 +2320,15 @@ void cClientHandle::SendHealth(void)
 
 
 
+void cClientHandle::SendHideTitle(void)
```
Maybe remove this method and if SendSetSubtitle() is called with a empty string, hide the title.

Howaner replied:
Oh, no. Its better to remove the SendResetTitle() and call it if SendSubtitle() is called with an empty string.

madmaxoft replied:
There is a slight possibility that you'd want to display the subtitle, but hide the main title, that's why a separate method is still a good idea.

Howaner commented on src/ClientHandle.cpp:
```diff
@@ -2568,6 +2586,42 @@ void cClientHandle::SendDisplayObjective(const AString & a_Objective, cScoreboar
 
 
 
+void cClientHandle::SendSetSubTitle(const cCompositeChat & a_SubTitle)
```
I don't like this name. Why not ```SendSubtitle(...)``|

madmaxoft replied:
To be consistent with `SendSetTitle()` and `SendHideTitle()`

NiLSPACE replied:
Would `SendShowTitle` not make more sense then?

madmaxoft replied:
Not really, because the packet doesn't show the title, it only sets the title and you need to send TitleTimes to actually show the title.

scottwillmoore replied:
That is actually incorrect. I managed to export the functions to the lua API and I was testing some of them the other day. It appears that `SendSetTitle(...)` and `SendSetSubTitle(...)` both when called (doesn't matter which one) will display the specifed type. `SendTitleTimes(...)` is just a preliminary step to displaying the title or subtitle. If both the title and the subtitle packet are sent at the same time they will both appear together.

bearbin commented:
@scottwillmoore Any update?

scottwillmoore commented:
Yeah, sorry for being so late, I have just been very busy lately and the laptop I was developing just recently broke and I have yet to replace.

The actual code appears to work and function how it should. When testing the functions in lua they all appeared to work and display the title correctly, however further testing should be conducted as I am just trying to recall this from memory.

At this stage the backend (C++) code is basically finished from my understanding, unless there are changes to the alogrithm or naming of the current functions that have been implemented. This could probably even be merged currently into the project at this stage, or into separate feature branch as it is functional code and works as expected.

A lua plugin still needs to be created which allows users to use commands to use this functionality ingame following the current syntax for the command as seen on the [wiki](http://minecraft.gamepedia.com/Commands#title).

Unfortunately I will not be currently able to work on the rest of the features until I have more time and my laptop has been repaired (hopefully under warranty!) Is anyone else keen on working on the lua plugin?

bearbin commented:
Well then I think this can be merged. @madmaxoft ?

madmaxoft commented:
Okay, merging.
@scottwillmoore could you please add the Lua API export as a separate PR? Thanks.
